### ISSUE TITLE
jwt_auth filter does not have full support of x509 public key 

### ISSUE THREAD

**Issue by JimmyCYJ** [istio/proxy#1295]
By reading code, two issues are observed which block jwt_auth filter from fully support of public key in x509 . 

1. We have Pubkeys::CreateFrom(const std::string &pkey, Type type) which takes public key in string and create Pubkeys object ([link](https://github.com/istio/proxy/blob/71f918875f883c07e6c2258e472de7fee1fb305d/src/envoy/http/jwt_auth/jwt.cc#L575)), but we only call this function when we fetch public keys in JWKS ([link](https://github.com/istio/proxy/blob/da18323bf354c55dac27a97a8dfbaa998b7c6a4d/src/envoy/http/jwt_auth/pubkey_cache.h#L74)). 

2. void Pubkeys::CreateFromPemCore(const std::string &pkey_pem) assumes that pkey_pem only contains one key ([link](https://github.com/istio/proxy/blob/71f918875f883c07e6c2258e472de7fee1fb305d/src/envoy/http/jwt_auth/jwt.cc#L457)).

**Comment by JimmyCYJ** [comment]
cc @diemtvu @liminw

**Comment by diemtvu** [comment]
FYI, I spent sometime to change CreateFromPemCore to parse pkey_pem (into json object) and create key (EvpPkeyFromStr) for each element. It doesn't work, with PEM_PUBKEY_BAD_BASE64 error (can be fixed by removing cert header and footer, and all newline charactor), then PEM_PUBKEY_PARSE_ERROR, which I stopped from debugging further. It all appears to me that implementation for x509 is not completed.

Plus, there is no clue how the base64 for the test kPublicKey was generated, but it cannot be parsed using online ssl checker.

**Comment by diemtvu** [comment]
+ @lizan, @qiwzhang do you have any insight on this?

**Comment by lizan** [comment]
Yes, we do not "fully" support x509 because we didn't need it, feel free to fix them.

More specifically 1. is for public key cache, we didn't intend to support loading PEM from remote URL, but only JWKS at the moment.

**Comment by liminw** [comment]
I think we need to prioritize this work. x509 is a very commonly used public key format. If we do not support x509 key format, we cannot claim that JWT authentication support is done.

**Comment by qiwzhang** [comment]
OK,  we will fix it then.  But not at istio/proxy repo.  I am in the middle of porting code to Envoy.  We should wait  for this PR https://github.com/envoyproxy/envoy/pull/2866
and fix it over there.

**Comment by liminw** [comment]
That plan sounds good to me.

**Comment by lizan** [comment]
Full X.509 spec is huge, we just need to support what is expected and demanded by JWT authentication and clarify it. With most of public authorization servers (Auth0, Google ID token, Firebase, Fb, etc), they only need JWKS.

**Comment by liminw** [comment]
All Google service account signed JWT tokens are using x509 key format.

**Comment by lizan** [comment]
You can always use JWK as well:
https://cloud.google.com/iam/docs/creating-managing-service-account-keys

`https://www.googleapis.com/service_accounts/v1/jwk/[SERVICE_ACCOUNT]@[PROJECT_ID].iam.gserviceaccount.com `

**Comment by lizan** [comment]
I'm not against adding support for more X.509 specs and PEM format, we might need X.509 certificate chain in JWT/JWK in its `x5c` or `x5u` claims in not far future. Though it will be substantial work to support them. Let's prioritize when we have a concrete use case for them.
